### PR TITLE
bugfix/38

### LINKED ISSUE #38: Poor Localizer Interception
Aircraft intercepting the localizer is very clunky; the turns are not smooth, and frequently overcompensate. The math for the turn calculations needs to be improved and simplified (so it is more clear how it works).

Also see zlsa/atc#711, n8rzz/atc#66, n8rzz/atc#86

Originally reported in zlsa/atc#717.

### LINKED ISSUE #419: No event recorded for intercepting ILS above localizer
Even if aircraft join the localizer above the glideslope, no penalty is issued, and no red response is shown in the log.

### PR DESCRIPTION
**Pull request by erikquinn**
Resolves #38.
Resolves #419.

Fix issues with localizer interception

Big thanks for @paul-at for the work on this!

### CODE REVIEW

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 1]
```diff
@@ -1393,8 +1395,26 @@ export default class AircraftModel {
                 break;
 
             case FLIGHT_PHASE.APPROACH: {
-                if (this.positionModel.distanceToPosition(this.mcp.nav1Datum) < AIRPORT_CONSTANTS.FINAL_APPROACH_FIX_DISTANCE_NM) {
-                    this.setFlightPhase(FLIGHT_PHASE.LANDING);
+                const datum = this.mcp.nav1Datum;
+                if (this.positionModel.distanceToPosition(datum) < AIRPORT_CONSTANTS.FINAL_APPROACH_FIX_DISTANCE_NM) {
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: No if/else here, early return

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 2]
```diff
@@ -1393,8 +1395,26 @@ export default class AircraftModel {
                 break;
 
             case FLIGHT_PHASE.APPROACH: {
-                if (this.positionModel.distanceToPosition(this.mcp.nav1Datum) < AIRPORT_CONSTANTS.FINAL_APPROACH_FIX_DISTANCE_NM) {
-                    this.setFlightPhase(FLIGHT_PHASE.LANDING);
+                const datum = this.mcp.nav1Datum;
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: Missing empty line below

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 3]
```diff
@@ -1393,8 +1395,26 @@ export default class AircraftModel {
...
+                if (this.positionModel.distanceToPosition(datum) < AIRPORT_CONSTANTS.FINAL_APPROACH_FIX_DISTANCE_NM) {
+                    const angleAwayFromLocalizer = this.mcp.course - this.positionModel.bearingToPosition(datum);
+                    const course = this.mcp.course;
+                    const courseOffset = getOffset(this, datum.relativePosition, course);
+                    const lateralDistanceFromCourseNm = nm(courseOffset[0]);
+
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: Remove empty line

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 4]
```diff
@@ -1643,21 +1661,23 @@ export default class AircraftModel {
      */
     updateLandingFailedLanding() {
         // Failed Approach
-        if ((this.approachDistance > 0.100) && (!this.projected)) {
+        if (!this.projected) {
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: Reverse logic and early return

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 5]
```diff
@@ -1393,8 +1395,26 @@ export default class AircraftModel {
...
+                    const angleAwayFromLocalizer = this.mcp.course - this.positionModel.bearingToPosition(datum);
+                    const course = this.mcp.course;
+                    const courseOffset = getOffset(this, datum.relativePosition, course);
+                    const lateralDistanceFromCourseNm = nm(courseOffset[0]);
+
+                    const isAlignedWithCourse = abs(lateralDistanceFromCourseNm) <= PERFORMANCE.MAXIMUM_DISTANCE_CONSIDERED_ESTABLISHED_ON_APPROACH_COURSE_NM;
+                    const isCourseStabilised = abs(angleAwayFromLocalizer) < PERFORMANCE.MAXIMUM_ANGLE_CONSIDERED_ESTABLISHED_ON_APPROACH_COURSE;
+                    const isOnGlideslope = abs(this.altitude - this.target.altitude) < 100;
+
+                    if (isCourseStabilised && isOnGlideslope && isAlignedWithCourse) {
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: Move this inner if/else to a new method and remove the else within that method.

File: src/assets/scripts/client/aircraft/AircraftModel.js [hunk 6]
```diff
@@ -1393,8 +1395,26 @@ export default class AircraftModel {
...
+                        // stabilised approach, switch to landing
+                        this.setFlightPhase(FLIGHT_PHASE.LANDING);
+                    } else {
+                        const runwayModel = this.fms.arrivalRunwayModel;
+                        if (this.altitude < runwayModel.elevation + PERFORMANCE.INSTRUMENT_APPROACH_MINIMUM_DESCENT_ALTITUDE) {
```
Comment (n8rzz) [inline src/assets/scripts/client/aircraft/AircraftModel.js]: Missing empty line above if

**Comment by erikquinn** [comment]
@paul-at On the practical test, they still spin _sometimes_, but it seems to be nearly completely fixed. Only seeing the issue when they're being cleared super, super late. :+1:

**Comment by paul-at** [comment]
@erikquinn kept this as a feature as I don't think ILS approaches can happen on such a short distance in real life. Should result in missed approach. If doesn't - happy to investigate further.

**Comment by erikquinn** [comment]
@paul-at right, but for runway 36 for example, eastbound on the base... When almost right on the localizer, giving left 330 and clearing is perfectly acceptable, and I don't remember anymore if that had the same issue.

I agree that giving 030 and clearing when they've got like 10 seconds to the localizer is crummy lol though even still, I think it's best to have it take it anyway. The spinning just derails everything when in reality the pilot would say something instead of doing that goofy maneuver. :wink:

File: src/assets/scripts/client/aircraft/AircraftModel.js (hunk 5 above)
Comment (erikquinn) [reply src/assets/scripts/client/aircraft/AircraftModel.js]: - `const isAlignedWithCourse` is already covered by a method `RunwayModel.isOnApproachCourse()`
`const isCourseStabilized` is already covered by a method `RunwayModel.isOnCorrectApproachHeading()`

Reworked this whole thing to make use of functions like those (and add ones we should have had), etc. This whole section reads really nicely now I think.

File: src/assets/scripts/client/aircraft/AircraftModel.js (hunk 4 above)
Comment (erikquinn) [reply src/assets/scripts/client/aircraft/AircraftModel.js]: done

File: src/assets/scripts/client/aircraft/AircraftModel.js (hunk 3 above)
Comment (erikquinn) [reply src/assets/scripts/client/aircraft/AircraftModel.js]: done

File: src/assets/scripts/client/aircraft/AircraftModel.js (hunk 2 above)
Comment (erikquinn) [reply src/assets/scripts/client/aircraft/AircraftModel.js]: done

File: src/assets/scripts/client/aircraft/AircraftModel.js (hunk 1 above)
Comment (erikquinn) [reply src/assets/scripts/client/aircraft/AircraftModel.js]: done

File: src/assets/scripts/client/constants/aircraftConstants.js [hunk 7]
```diff
@@ -132,6 +150,16 @@ export const PERFORMANCE = {
      */
     LANDING_ASSIGNED_SPEED_DISTANCE_NM: 5,
 
+    /**
+     * Maximum vertical distance between the aircraft and the glidepath to
+     * consider the aircraft to be "established on the localizer"
```
Comment (erikquinn) [inline src/assets/scripts/client/constants/aircraftConstants.js]: on the GLIDESLOPE, dummy

Comment (n8rzz) [reply src/assets/scripts/client/constants/aircraftConstants.js]: is that `dummy` pointed at you, as in _you're the dummy_?

Comment (erikquinn) [reply src/assets/scripts/client/constants/aircraftConstants.js]: I am the dummy, yes :smiley:

**Pull request merged**